tower-http/src/set_response_header.rs

```diff
@@ -0,0 +1,457 @@
+//! #
+//! # Ok(())
+//! # }
+//! ```
+
+use futures_util::ready;
+use http::{header::HeaderName, HeaderValue, Response};
+use pin_project::pin_project;
+use std::{
+    fmt,
+    pin::Pin,
+    task::{Context, Poll},
+};
+use std::{future::Future, marker::PhantomData};
+use tower_layer::Layer;
+use tower_service::Service;
+
+/// Layer that applies [`SetResponseHeader`] which adds a response header.
+///
+/// See [`SetResponseHeader`] for more details.
+pub struct SetResponseHeaderLayer<M, T> {
+    header_name: HeaderName,
+    make: M,
+    mode: InsertHeaderMode,
+    _marker: PhantomData<fn() -> T>,
+}
+
+impl<M, T> fmt::Debug for SetResponseHeaderLayer<M, T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("SetResponseHeaderLayer")
+            .field("header_name", &self.header_name)
+            .field("mode", &self.mode)
+            .field("make", &std::any::type_name::<M>())
+            .finish()
+    }
+}
+
+impl<M, T> SetResponseHeaderLayer<M, T> {
+    /// Create a new [`SetResponseHeaderLayer`].
+    pub fn new(header_name: HeaderName, make: M) -> Self
+    where
+        M: MakeHeaderValue<T>,
+    {
+        Self {
+            make,
+            header_name,
+            mode: InsertHeaderMode::OverrideExisting,
+            _marker: PhantomData,
+        }
+    }
+
+    /// Set which mode to use when inserting the header.
+    ///
+    /// Defaults to [`InsertHeaderMode::OverrideExisting`].
+    pub fn mode(mut self, mode: InsertHeaderMode) -> Self {
+        self.mode = mode;
+        self
+    }
+}
+
+/// The mode to use when inserting a header into a request or response.
+#[derive(Debug, Clone)]
+#[non_exhaustive]
+pub enum InsertHeaderMode {
+    /// Insert the header, overriding any previous values the header might have.
+    OverrideExisting,
+    /// Append the header and keep any previous values.
+    Append,
+    /// Insert the header only if it is not already present.
+    SkipIfPresent,
+}
+
+impl<T, S, M> Layer<S> for SetResponseHeaderLayer<M, T>
+where
+    M: MakeHeaderValue<T> + Clone,
+{
+    type Service = SetResponseHeader<S, M>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        SetResponseHeader {
+            inner,
+            header_name: self.header_name.clone(),
+            make: self.make.clone(),
+            mode: self.mode.clone(),
+        }
+    }
+}
+
+impl<M, T> Clone for SetResponseHeaderLayer<M, T>
+where
+    M: Clone,
+{
+    fn clone(&self) -> Self {
+        Self {
+            make: self.make.clone(),
+            header_name: self.header_name.clone(),
+            mode: self.mode.clone(),
+            _marker: PhantomData,
+        }
+    }
+}
+
+/// Middleware that sets a header on the response.
+#[derive(Clone)]
+pub struct SetResponseHeader<S, M> {
+    inner: S,
+    header_name: HeaderName,
+    make: M,
+    mode: InsertHeaderMode,
+}
+
+impl<S, M> SetResponseHeader<S, M> {
+    /// Create a new [`SetResponseHeader`].
```

Comment:
This should probably say something like
```suggestion
    /// Create a new [`SetResponseHeader`].
    ///
    /// By default, the layer will construct services configured with
    /// [`InsertHeaderMode::OverrideExisting`]. This will replace any
    /// previously set values for that header. This behavior can be
    /// changed using the [`mode`] method.
```